# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [60]:
import pandas as pd

temperatures = pd.Series(
    [72, 75, 68, 71, 73],  
    index=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"],  
    name="Daily High Temperatures (°F)"
)

print("Temperature Series:")
print(temperatures)

print("\nData type of Series values:")
print(temperatures.dtype)

Temperature Series:
Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
Name: Daily High Temperatures (°F), dtype: int64

Data type of Series values:
int64


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [61]:
product_data = {
    "product": ["Widget", "Gadget", "Doohickey"],
    "price": [19.99, 24.99, 12.49],
    "quantity": [100, 75, 150]
}

products = pd.DataFrame(product_data)

print("Products DataFrame:")
print(products)

print("\nShape of the DataFrame:")
print(products.shape)



Products DataFrame:
     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150

Shape of the DataFrame:
(3, 3)


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [62]:
price_series = products["price"]
print("Price Column (as a Series):")
print(price_series)

product_quantity_df = products[["product", "quantity"]]
print("\nProduct and Quantity Columns (as a DataFrame):")
print(product_quantity_df)

Price Column (as a Series):
0    19.99
1    24.99
2    12.49
Name: price, dtype: float64

Product and Quantity Columns (as a DataFrame):
     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

#### Answers
1. When a DataFrame is constructed as a dictionary, it has keys to set as the column names, and the values (usually lists) become the data entries for each respective column. This structure enables Pandas to distribute data.
2. `df['price']` generates a series, so it selects a single column and produces a one dimiensional data structure. On the other hand, `df[['price']]` produces a DataFrame, this is because it uses a list to specify columns in two dimensional tabular structure.
3. A custom index, like the names of the days, has a specific purpose for contextual meaning, for example, dates, more product labels, or other group names. This makes the dataset more readable and makes the data easier to access and analyze compared to using the default numeric index. 


### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [63]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('../data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)


In [64]:
df_raw = pd.read_csv(csv_path)

print("Preview of Raw Data:")
display(df_raw.head())

print("\nShape of the dataset (rows, columns):")
print(df_raw.shape)

print("\nColumn Information:")
print(df_raw.dtypes)

print("\nGeneral Information:")
df_raw.info()

print("\nSummary Statistics (Numeric Columns):")
display(df_raw.describe())

print("\nPossible datetime or categorical columns:")
for col in df_raw.columns:
    if 'start' in col.lower() or 'end' in col.lower() or 'type' in col.lower():
        print(f"- {col}")

print("\nDetailed Memory Usage (in MB):")
memory_usage_MB = df_raw.memory_usage(deep=True).sum() / (1024 ** 2)
print(f"Total memory used: {memory_usage_MB:.2f} MB")


Preview of Raw Data:


ride_id  rideable_type               started_at  \
0  67BB74BD7667BAB7  electric_bike  2024-09-30 23:12:01.622   
1  5AF1AC3BA86ED58C  electric_bike  2024-09-30 23:19:25.409   
2  7961DD2FC1280CDC   classic_bike  2024-09-30 23:32:24.672   
3  2E16892DEEF4CC19   classic_bike  2024-09-30 23:42:11.207   
4  AAF0220F819BEE01  electric_bike  2024-09-30 23:49:25.380   

                  ended_at         start_station_name start_station_id  \
0  2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave           bdd4c3   
1  2024-10-01 00:42:09.933                        NaN              NaN   
2  2024-10-01 00:23:18.647     St. Clair St & Erie St           9c619a   
3  2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave           72a04d   
4  2024-10-01 00:06:27.476          900 W Harrison St           11da85   

           end_station_name end_station_id  start_lat  start_lng    end_lat  \
0                       NaN            NaN  42.012342 -87.688243  41.970000   
1    Benson Ave & Church St         a10cf0  42.070000 -87.730000  42.048214   
2  LaSalle St & Illinois St         fbd1ad  41.894345 -87.622798  41.890762   
3    Loomis St & Archer Ave         896337  41.895954 -87.667728  41.841633   
4         900 W Harrison St         11da85  41.874754 -87.649807  41.874754   

     end_lng member_casual  
0 -87.650000        casual  
1 -87.683485        casual  
2 -87.631697        member  
3 -87.657435        casual  
4 -87.649807        member


Shape of the dataset (rows, columns):
(100000, 13)

Column Information:
ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object

General Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             100000 non-null  object 
 1   rideable_type       100000 non-null  object 
 2   started_at          100000 non-null  object 
 3   ended_at            100000 non-null  object 
 4   start_station_name  89623 non-null   object 
 5   start_station_id    89623 non-null  

start_lat      start_lng       end_lat       end_lng
count  100000.000000  100000.000000  99913.000000  99913.000000
mean       41.898817     -87.644839     41.899246    -87.645279
std         0.045897       0.027118      0.046581      0.028055
min        41.648501     -87.840000     41.610000    -89.120000
25%        41.879356     -87.658902     41.879472    -87.659172
50%        41.894666     -87.641180     41.894822    -87.641255
75%        41.925566     -87.627716     41.925858    -87.628579
max        42.070000     -87.530000     43.930000    -86.050000


Possible datetime or categorical columns:
- rideable_type
- started_at
- ended_at
- start_station_name
- start_station_id
- end_station_name
- end_station_id
- start_lat
- start_lng
- end_lat
- end_lng

Detailed Memory Usage (in MB):
Total memory used: 63.70 MB


#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [65]:
import pandas as pd
from pathlib import Path

csv_path = Path("../data/202410-divvy-tripdata-100k.csv")

categorical_cols = [
    "rideable_type",
    "start_station_name",
    "start_station_id",
    "end_station_name",
    "end_station_id",
    "member_casual"
]

rides = pd.read_csv(
    csv_path,
    dtype={col: "category" for col in categorical_cols},  
    parse_dates=["started_at", "ended_at"]                
)

print("Dataset Info After Type Optimization:")
rides.info()

optimized_memory_MB = rides.memory_usage(deep=True).sum() / (1024 ** 2)
print(f"\nOptimized total memory used: {optimized_memory_MB:.2f} MB")

previous_memory_MB = 63.70  
reduction = (previous_memory_MB - optimized_memory_MB) / previous_memory_MB * 100
print(f"Memory reduction: {reduction:.2f}%")



Dataset Info After Type Optimization:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  object        
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   category      
 5   start_station_id    89623 non-null   category      
 6   end_station_name    89485 non-null   category      
 7   end_station_id      89485 non-null   category      
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  category    

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [66]:
start_min = rides["started_at"].min()
start_max = rides["started_at"].max()

print("Date Range of Rides:")
print(f"- Earliest start date: {start_min}")
print(f"- Latest start date:   {start_max}")

missing_counts = rides.isna().sum()

missing_df = missing_counts.to_frame(name="count")
missing_df["percentage"] = (missing_df["count"] / len(rides)) * 100

print("\nMissing Data Summary:")
display(missing_df.sort_values(by="percentage", ascending=False))



Date Range of Rides:
- Earliest start date: 2024-09-30 23:12:01.622000
- Latest start date:   2024-10-31 23:54:02.851000

Missing Data Summary:


count  percentage
end_station_name    10515      10.515
end_station_id      10515      10.515
start_station_name  10377      10.377
start_station_id    10377      10.377
end_lat                87       0.087
end_lng                87       0.087
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_lat               0       0.000
start_lng               0       0.000
member_casual           0       0.000

#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [67]:
rides["trip_duration_min"] = (rides["ended_at"] - rides["started_at"]).dt.total_seconds() / 60

print("Trip Duration (minutes) - Summary Statistics:")
display(rides["trip_duration_min"].describe())

print("\nSample Rows (Started, Ended, Duration):")
display(rides[["started_at", "ended_at", "trip_duration_min"]].head())

rides.set_index("started_at", inplace=True)

print("\nNew DataFrame Index:")
print(rides.index)

print("\nFirst few rows after setting index:")
display(rides.head())


Trip Duration (minutes) - Summary Statistics:


count    100000.000000
mean         16.144576
std          52.922539
min           0.006533
25%           5.489271
50%           9.423592
75%          16.407171
max        1499.949717
Name: trip_duration_min, dtype: float64


Sample Rows (Started, Ended, Duration):


started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583
3 2024-09-30 23:42:11.207 2024-10-01 00:10:16.831          28.093733
4 2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          17.034933


New DataFrame Index:
DatetimeIndex(['2024-09-30 23:12:01.622000', '2024-09-30 23:19:25.409000',
               '2024-09-30 23:32:24.672000', '2024-09-30 23:42:11.207000',
               '2024-09-30 23:49:25.380000', '2024-09-30 23:49:40.016000',
               '2024-10-01 00:00:53.414000', '2024-10-01 00:05:44.954000',
               '2024-10-01 00:06:12.035000', '2024-10-01 00:10:03.646000',
               ...
               '2024-10-31 23:36:04.200000', '2024-10-31 23:36:34.956000',
               '2024-10-31 23:36:49.500000', '2024-10-31 23:38:20.262000',
               '2024-10-31 23:44:03.832000', '2024-10-31 23:44:23.211000',
               '2024-10-31 23:44:45.948000', '2024-10-31 23:50:31.160000',
               '2024-10-31 23:53:02.355000', '2024-10-31 23:54:02.851000'],
              dtype='datetime64[ns]', name='started_at', length=100000, freq=None)

First few rows after setting index:


ride_id  rideable_type  \
started_at                                                 
2024-09-30 23:12:01.622  67BB74BD7667BAB7  electric_bike   
2024-09-30 23:19:25.409  5AF1AC3BA86ED58C  electric_bike   
2024-09-30 23:32:24.672  7961DD2FC1280CDC   classic_bike   
2024-09-30 23:42:11.207  2E16892DEEF4CC19   classic_bike   
2024-09-30 23:49:25.380  AAF0220F819BEE01  electric_bike   

                                       ended_at         start_station_name  \
started_at                                                                   
2024-09-30 23:12:01.622 2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave   
2024-09-30 23:19:25.409 2024-10-01 00:42:09.933                        NaN   
2024-09-30 23:32:24.672 2024-10-01 00:23:18.647     St. Clair St & Erie St   
2024-09-30 23:42:11.207 2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave   
2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          900 W Harrison St   

                        start_station_id          end_station_name  \
started_at                                                           
2024-09-30 23:12:01.622           bdd4c3                       NaN   
2024-09-30 23:19:25.409              NaN    Benson Ave & Church St   
2024-09-30 23:32:24.672           9c619a  LaSalle St & Illinois St   
2024-09-30 23:42:11.207           72a04d    Loomis St & Archer Ave   
2024-09-30 23:49:25.380           11da85         900 W Harrison St   

                        end_station_id  start_lat  start_lng    end_lat  \
started_at                                                                
2024-09-30 23:12:01.622            NaN  42.012342 -87.688243  41.970000   
2024-09-30 23:19:25.409         a10cf0  42.070000 -87.730000  42.048214   
2024-09-30 23:32:24.672         fbd1ad  41.894345 -87.622798  41.890762   
2024-09-30 23:42:11.207         896337  41.895954 -87.667728  41.841633   
2024-09-30 23:49:25.380         11da85  41.874754 -87.649807  41.874754   

                           end_lng member_casual  trip_duration_min  
started_at                                                           
2024-09-30 23:12:01.622 -87.650000        casual          67.984200  
2024-09-30 23:19:25.409 -87.683485        casual          82.742067  
2024-09-30 23:32:24.672 -87.631697        member          50.899583  
2024-09-30 23:42:11.207 -87.657435        casual          28.093733  
2024-09-30 23:49:25.380 -87.649807        member          17.034933

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

#### Task 2 – Interpretation

1. Panda automatically assigned to object data type for both `started_at` and `member_casual`. This default might be problematic because the `started_at` column should use the datetime data type to have accurate time calculations, and for `member_casual` column should be stored as a categorical type. Converting this column type to their right types improves performance, reduces memory usage, and enables time-based or categorical analysis.
2. As shown, the station IDs appear to be hashed, and it is similar to Git commit hashes, because the length of the alphanumeric format of the ride IDs, so I think the station IDs are derived based on the station's name or geographic coordinates to create a unique fingerprint for each trip.
3. Method changing refers to performing multiple functions or methods in sequence within a single command for more readability and efficiency. For `df.isna()`, it generates a Boolean Dataframe if there missing value, it shows `TRUE`  and `False` otherwise. For the `.sum()`, it counts the total number of TRUE values in each column and gives the total number of entries in each column.
4. The 10% of missing values in station columns and 0% in coordinates show that many rides begin or end at an unregistered, temporary, or decommissioned station, so the GPS didn't have a matching station name or ID. To handle this issue, if the coordinates are valid, you could map the missing station names and use the nearest known station, or if the missingness is widespread or systematic, flag them as 'Unknown Station' to preserve the data while acknowledging uncertainty.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

#### Answers
1. Comparison of Memory Usage

   As calculated in Task 2a.3, the Usage of the memory was approximately 63.7 MB, and then after optimizing the data types in Task 2b.3, the same dataset required only 12.85 MB, which represents a reduction of approximately 79.8% in total memory usage. This reduction is significant due to the change from the objective type (default) to the categorical types, which is the right type for this dataset, and also changing the datetime data type, which uses a fixed-width numeric representation instead of variable-length strings. However, the memory numbers are different from what is shown at the bottom of `df.info()`, and this is because the `df.info()` method makes a shallow memory estimate without counting the overhead and internal Python object overhead of string and categorical data. The `memory_usage(deep=True).sum()` method gives a more accurate estimate of true memory usage by considering all underlying object references and string data. When scaling or analyzing data by `df.memory_usage(deep=True).sum()` gives the most accurate estimate, as it shows the total in-memory footprint of the DataFrame.

2. Estimating Safe Dataset Sizes on a 16GB System

   System memory totals 16GB, we reserve = 30% of memory for the operating system, so 11.2GB is left for data analysis.

   Operations in Panda require, on average, 5-10 times the size of a dataset in RAM. Following these assumptions,

  - Optimistic scenario (5x multiplier): 11.2GB/5 = 2.24GB

   - Conservatively (10x multiplier): 11.2GB/10 = 1.12GB

   Therefore, on a 16GB system, the approximate safe working range for the size of the dataset is 1.1GB - 2.2GB more complicated workloads, more estimation will be required. 
 For best system performance and more reliable interactivity in a notebook, is it better to work withen 1GB - 1.5GB datasets. This ensures RAM will be available for visualization, computation, and intermediate DataFrame creation.


### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [68]:
member_electric = rides[(rides['member_casual'] == 'member') & 
                        (rides['rideable_type'] == 'electric_bike')]

num_member_electric = len(member_electric)

percent_member_electric = (num_member_electric / len(rides)) * 100

stations_of_interest = [
    "Streeter Dr & Grand Ave",
    "DuSable Lake Shore Dr & Monroe St",
    "Kingsbury St & Kinzie St"
]

trips_selected_stations = rides[rides['start_station_name'].isin(stations_of_interest)]
num_selected_stations = len(trips_selected_stations)

print("Number of trips by members using electric bikes:", num_member_electric)
print(f"Percentage of all trips: {percent_member_electric:.2f}%")
print("Number of trips starting at selected stations:", num_selected_stations)


Number of trips by members using electric bikes: 33121
Percentage of all trips: 33.12%
Number of trips starting at selected stations: 2702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [69]:
rides['is_weekend'] = rides.index.dayofweek >= 5
start_hour = rides.index.hour

rides['time_of_day'] = "Midday"

morning_mask = start_hour.isin([7, 8, 9])
rides.loc[morning_mask, 'time_of_day'] = "Morning Rush"

evening_mask = start_hour.isin([16, 17, 18])
rides.loc[evening_mask, 'time_of_day'] = "Evening Rush"
rides['time_of_day'].value_counts()



time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64

#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [70]:
rides['duration_min'] = (rides['ended_at'] - rides.index).dt.total_seconds() / 60
iltered_trips = rides.query(
    "member_casual == 'casual' and is_weekend == True and duration_min > 20 and rideable_type == 'electric_bike'"
)

num_filtered = len(filtered_trips)

percent_filtered = (num_filtered / len(rides)) * 100

avg_duration = filtered_trips['duration_min'].mean()

print("Number of matching trips:", num_filtered)
print(f"Percentage of all trips: {percent_filtered:.2f}%")
print(f"Average duration of these trips: {avg_duration:.2f} minutes")


Number of matching trips: 1501
Percentage of all trips: 1.50%
Average duration of these trips: 40.37 minutes


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [71]:
rides.iloc[:10, rides.columns.get_indexer(['member_casual', 'rideable_type', 'duration_min'])]

rides.loc['2024-10-15':'2024-10-17', ['member_casual', 'rideable_type', 'duration_min']]

num_trips_range = rides.loc['2024-10-15':'2024-10-17'].shape[0]
print("Number of trips from Oct 15–17, 2024:", num_trips_range)


Number of trips from Oct 15–17, 2024: 7235


#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

1. Advantages of `isin()`

   Using the `isin()` method can be more efficient and scalable than multiple logical conditions, such as
   `(col == 'A') | (col == 'B') | (col == 'C')`
   For this method, the most important benefit is the efficiency and flexibility, so you can pass an entire list of values directly into `isin()`. This method leads to clearer and simpler code, reduces errors, and makes names or filtering criteria easier to adjust.

2. Conditional assignment order in Task 3b

   By initializing the `time_of_day` column to "Midday", all records have a valid default value before more specific categories are assigned, such as "Morning Rush" and "Evening Rush". This order shows missing values  `(NaN)` for records that do not meet the rush hour criteria. When records are assigned values in a different order or left without a default, unintended overwriting or lack of definition occurs, resulting in inaccurate classifications.

3. `query()` vs. Boolean indexing

   When it comes to multiple logical conditions and complex filtering, the `query()` method is preferred for its readability and conciseness. Its SQL syntax is perfect for exploratory analysis and presentation. However, using Boolean indexing can be more articulate and is more appropriate in cases requiring the use of Python variables, manipulating column names with whitespace or special symbols. Boolean indexing provides more transparency and debugging control in data processing workflows.

#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

#### Answer
Pandas have multiple indexing methods, such as bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing(`iloc[]`). Even though brackets can be used to select a column and slice rows `df[0:5]`, it is not recommended for row operations ambiguous and context-dependent behavior. The best practice way is to use brackets only for column selection, where its meaning while using brackets is constant and clear, for instance (`df['column_name'])` or `df[['col1', 'col2']]`. Using `loc[]` for label-based selection to access rows and columns using index labels.  Using `iloc[]` for position-based selection, while referring to rows or columns through numerical indices. So, using `df.loc[]` and `df.iloc` makes your intent clear to both Python and other readers, reducing logical errors when the dataset changes.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [72]:
rides['hour'] = rides.index.hour

hour_counts = rides['hour'].value_counts()
peak_hour = hour_counts.idxmax()
peak_hour_trips = hour_counts.max()

print(f"Most popular hour for trips: {peak_hour}:00")
print(f"Number of trips during that hour: {peak_hour_trips}")

rides['day_name'] = rides.index.day_name()

day_counts = rides['day_name'].value_counts()
busiest_day = day_counts.idxmax()
busiest_day_trips = day_counts.max()

print(f"Busiest day of the week: {busiest_day}")
print(f"Number of trips on that day: {busiest_day_trips}")

ordered_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
day_counts = day_counts.reindex(ordered_days)

print("\nTrips by day of the week (Monday–Sunday):")
print(day_counts)



Most popular hour for trips: 17:00
Number of trips during that hour: 10574
Busiest day of the week: Wednesday
Number of trips on that day: 16513

Trips by day of the week (Monday–Sunday):
day_name
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64


#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [73]:
avg_duration_by_rider = rides.groupby('member_casual', observed=False)['duration_min'].mean()

avg_duration_by_bike = rides.groupby('rideable_type', observed=False)['duration_min'].mean()

trip_count_groupby = rides.groupby('member_casual', observed=False).size()

trip_count_value_counts = rides['member_casual'].value_counts()

print("Average trip duration by rider type (minutes):")
print(avg_duration_by_rider)
print("\nAverage trip duration by bike type (minutes):")
print(avg_duration_by_bike)
print("\nTrip count by rider type using groupby().size():")
print(trip_count_groupby)
print("\nTrip count by rider type using value_counts():")
print(trip_count_value_counts)


Average trip duration by rider type (minutes):
member_casual
casual    23.978046
member    11.984493
Name: duration_min, dtype: float64

Average trip duration by bike type (minutes):
rideable_type
classic_bike     20.337410
electric_bike    12.033618
Name: duration_min, dtype: float64

Trip count by rider type using groupby().size():
member_casual
casual    34686
member    65314
dtype: int64

Trip count by rider type using value_counts():
member_casual
member    65314
casual    34686
Name: count, dtype: int64


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [74]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Task 4c code here...

# 1. Filter for weekend electric bike trips
weekend_electric = rides[(rides['is_weekend'] == True) &
                         (rides['rideable_type'] == 'electric_bike')]

# Total number of weekend electric bike trips (before sampling)
total_weekend_electric = len(weekend_electric)

# 2. Use iloc[] to select the first 1000 trips from this filtered dataset
weekend_electric_sample = weekend_electric.iloc[:1000].copy()

# 3. Reset index so started_at becomes a column
weekend_electric_sample = weekend_electric_sample.reset_index()

# 4. Ensure we have a column named trip_duration_min for export
#    (alias from duration_min)
weekend_electric_sample['trip_duration_min'] = weekend_electric_sample['duration_min']

# 5. Export to CSV with the required columns
cols_to_export = ['started_at', 'ended_at', 'member_casual',
                  'trip_duration_min', 'time_of_day']

weekend_electric_sample[cols_to_export].to_csv(output_file, index=False)

# 6. Report how many total weekend electric bike trips existed before sampling
print(f"Total weekend electric bike trips (before sampling): {total_weekend_electric}")
print(f"Exported first {len(weekend_electric_sample)} trips to: {output_file}")



Total weekend electric bike trips (before sampling): 13026
Exported first 1000 trips to: output\weekend_electric_trips.csv


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

1. Conceptual model of `groupby()`
   The `groupby()` function in Pandas follows the "split-apply-combine" methodology. In **Split**, the data is separated into multiple groups using one or more key columns. In **Apply**, each group receives a computation or aggregation, for example, the groups (mean, size, or total sum), which is done independently. For **Combine** each group's results are brought back together into a single consolidated DataFrame or series. This process enables efficient and structured analysis of data subsets.

2. `value_counts()` vs. `groupby()`

   The `value_counts()` provides a quick and simple way to count the frequency of unique values in a single column of a data set. It is simplistic, fast, and works well for one-dimensional summaries.

   The `groupby()` method works with `.size()` or aggregation functions to provide more developed and complex operations, while each method gives the same result for counting a single column, only the `groupby ()` function can manage multi-level groupings or other statistical operations for each group. This makes `groupby()` a more advanced and analytical tool in these cases.

3. Exporting Managed Data Sets

   In Task 4c, the `reset_index()` function was applied prior to exporting the dataset to enable conversion of the datetime index (`started_at`) into a regular column. This ensures the start time data is included in the exported CSV. If the DataFrame were exported without resetting the index and with `index=False`, the datetime index would not appear in the file, resulting in the loss of critical time-based information. Resetting the index ensures exported datasets are intact, unambiguous, and prepared for subsequent analyses or visual presentation.


#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [75]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides
    .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18]))]
    .loc[(rides['start_station_name'].str.contains('Lake Shore|Lakefront', 
                                                    case=False, 
                                                    na=False))]
    .loc[rides['trip_duration_min'] > 15]
    .sort_values('trip_duration_min', ascending=False)
    .head(1000)
    .reset_index()
    [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
      'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

Found 310 long rush-hour trips from lakefront stations

CSV file size: 40.57 KB
Pickle file size: 55.21 KB
Size difference: 14.64 KB

Load time comparison:
CSV:
1.49 ms ± 122 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Pickle:
449 μs ± 21.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name          category
end_station_name            category
member_casual               category
rideable_type               category
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


### Answers

1. Method chaining:

   Using this pattrens makes the code more readable because each transformation is on its own line, like a pipeline. You can see the sequence of steps in order instead of trying to parse one long, crowded line. The parentheses remove the need for backslashes, so the syntax is cleaner and less error-prone. It also has the added advantage of making debugging easier: if something in the chain breaks, you can easily comment out or remove a single step in the chain, or even temporarily split the chain and inspect an intermediate result. As each method call is on its own line, it's easy to see which operation is causing the issue. This style works well with `git diff`. Each change affects only one line. However, in a very long chain, it becomes hard to read or debug, so in complex analysis, it is better to split into smaller steps.

2. Data typs

   Pickle preserved the datatime and category types that were lost when saving to CSV which converted them to plain object strings. This preservation is important because it keeps memory usage low, speeds up perations, and maintains correct data behavior.

3. Trade-offs

   Pickle loads faster and keeps all data types, making it ideal for Python/Pandas and repeated analysis. This provides a version with a cleaned Dataframe. CSV is slower and loses types, but remains universal, readable, and shareable through different tools. Use Pickle for performance and accuracy, and CSV for portability and collaboration. 



   

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

1. NumPy vs Pandas

   The biggest conceptual change from using NumPy's position based data to Pandas labeled, table like data. the hardest part was method changing, and I work through it by making smaller examples and using parentheses for clarity where needed.

2. Real data experince

   Real csv data required handling missing values, mixed types, and datetimes, unlike the clean NumPy arrays. Pandas proved more capable in cleaning, orgnizing, and analyzing real world datasets.

3. Learning & Application

   The most useful skill were data type managment and using `loc`,and `iloc`. I fell 7/10 prepared to use Pandas. more practice with large datasets would raise that score

4. Feedback

   Time spent: __12_ hours (breakdown optional)

   Most helpful part of the assignment: _Problem 3__

   One specific improvement suggestion: __Provide visual summaries (flowcharts or diagrams)_